### PR TITLE
make examples subpackage (fuse_examples)

### PR DESCRIPTION
**Pull request by mosheraboh**
Hi @alex-golts @YoelShoshan @SagiPolaczek !
After many issues,
I've changed the examples package to be be subpackage.
Please carefully review - cause we can change it only once :) .

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -145,7 +145,7 @@ Broadly speaking, the supported DL libraries are:
 * [pytorch-lightning](https://www.pytorchlightning.ai/)
 
 Before you ask - **pytorch-lightning and FuseMedML play along very nicely and have in practice orthogonal and additive benefits :)**
-See [Simple FuseMedML + PytorchLightning Example](./examples/fuse_examples/imaging/classification/mnist/run_mnist.py) for simple supervised learning cases, and [this example ](./examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py) for completely custom usage of pytorch-lightning and FuseMedML - useful for advanced scenarios such as Reinforcement Learning and generative models.
+See [Simple FuseMedML + PytorchLightning Example](./fuse_examples/imaging/classification/mnist/run_mnist.py) for simple supervised learning cases, and [this example ](./fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py) for completely custom usage of pytorch-lightning and FuseMedML - useful for advanced scenarios such as Reinforcement Learning and generative models.
```
Comment (SagiPolaczek) [inline README.md:148]: path should be:
`./fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py`
(diff file name)

File: fuse/dl/README.md [hunk 2]
```diff
@@ -9,7 +9,7 @@ FuseMedML is very flexible. This means that a user can choose to only use some c
 FuseMedML uses PyTorch Lightning as it's model training "manager".
 There are two main supported ways in which lightning is used in fuse:
 ### 1. Ready-made wrapper use (suitable mostly for supervised learning use-cases)
-(see [MNIST example](../../examples/fuse_examples/imaging/classification/mnist/run_mnist.py)).
+(see [MNIST example](../../fuse_examples/imaging/classification/mnist/run_mnist.py)).
```
Comment (SagiPolaczek) [inline fuse/dl/README.md:12]: here it should be
`simple_mnist_starter.py`

File: README.md (hunk 1 above)
Comment (SagiPolaczek) [reply README.md:148]: or with
`simple_mnist_starter.py`

File: fuse/dl/README.md [hunk 3]
```diff
@@ -41,7 +41,7 @@ The basic classes are:
 1. `LossBase` - A generic template class that can be used for implementing any loss function. All it does is receives a `batch_dict` and returns a loss tensor.
 2. `LossWarmUp` - A wrapper class that gets an existing loss function and zeros its value for a given number of iterations. This is useful in a multiple loss function scenario, where one loss should be stabilized before the other one comes into play.
 3. `LossDefault` - Default loss implementation of FuseMedML. It is still a generic class in the sense that is can operate on any actual loss function implementation. In addition, it can receive a scalar multiplier value, and a custom preprocessing function that operated on the `batch_dict` before the loss operates on it.
-[The MNIST example uses `LossDefault`](../../examples/fuse_examples/imaging/classification/mnist/run_mnist.py) as follows:
+[The MNIST example uses `LossDefault`](../../fuse_examples/imaging/classification/mnist/run_mnist.py) as follows:
```
Comment (SagiPolaczek) [inline fuse/dl/README.md:44]: `run_mnist_custom_pl_imp.py` or `simple_mnist_starter.py`

File: fuse/dl/README.md [hunk 4]
```diff
@@ -62,7 +62,7 @@ Note again, that while these are specific loss functions that are implemented, i
 This module contains DL model and architecture related classes. A FuseMedML model is very similar to a PyTorch model. The only difference is that in its `forward` function it gets as input a fuse `batch_dict`, extracts the input data from the input keys, forwards it through the model and writes the output to the `batch_dict`'s output keys.
 The most basic class is `ModelWrapSeqToDict`. It is FuseMedML's wrapper for PyTorch models to be used in fuse. It is initialized with an existing PyTorch model, and a list of input and output keys.
 Optionally, a pre and post processing function can be provided to be applied on the `batch_dict` before and after model run.
-[The MNIST example uses `ModelWrapSeqToDict`](../../examples/fuse_examples/imaging/classification/mnist/run_mnist.py) to wrap a PyTorch model as follows:
+[The MNIST example uses `ModelWrapSeqToDict`](../../fuse_examples/imaging/classification/mnist/run_mnist.py) to wrap a PyTorch model as follows:
```
Comment (SagiPolaczek) [inline fuse/dl/README.md:65]: `run_mnist_custom_pl_imp.py` or `simple_mnist_starter.py` (both using wrapper with LeNet)

**Review by SagiPolaczek — COMMENTED** [review]
we are using `run_mnist.py` where it should be:
`run_mnist_custom_pl_imp.py` or `simple_mnist_starter.py`

File: README.md [hunk 5]
```diff
@@ -202,28 +202,33 @@ $ pip install -e .[all]
 ```
 This mode installs all the currently publicly available domain extensions - fuseimg as of now, fusedrug will be added soon.
 
-In this mode you can also install and run our end to end examples using:
+To install `FuseMedML` with an included collection of examples install it using:
```
Comment (alex-golts) [inline README.md:205]: I would also include `pip install -e .[all]` to install in interactive mode but without examples

File: fuse_examples/multimodality/ehr_transformer/dataset.py [hunk 6]
```diff
@@ -16,7 +16,7 @@
 from fuse.data.utils.split import dataset_balanced_division_to_folds
 from fuse.data.utils.export import ExportDataset
 
-from examples.fuse_examples.multimodality.ehr_transformer.utils import (
+from fuse_examples.multimodality.ehr_transformer.utils import (
```
Comment (alex-golts) [inline fuse_examples/multimodality/ehr_transformer/dataset.py:19]: glad to see I'm not the only one who did it :-)

File: setup.py [hunk 7]
```diff
@@ -32,8 +32,17 @@
         if not line.startswith("#"):
             fuseimg_requirements.append(line.strip())
 
+# list of requirements for fuse_examples
+fuse_exampels_requirements = []
+with open(os.path.join(HERE, "fuse_examples/requirements.txt"), "r") as fh:
+    for line in fh:
+        if not line.startswith("#"):
+            fuseimg_requirements.append(line.strip())
```
Comment (alex-golts) [inline setup.py:40]: should be `fuse_examples_requirements.append` right?

File: setup.py [hunk 8]
```diff
@@ -32,8 +32,17 @@
         if not line.startswith("#"):
             fuseimg_requirements.append(line.strip())
 
+# list of requirements for fuse_examples
+fuse_exampels_requirements = []
```
Comment (alex-golts) [inline setup.py:36]: *examples (typo)

**Review by alex-golts — CHANGES_REQUESTED** [review]
Looks good! I think it makes more sense this way.
please see few small comments/corrections

**Review by SagiPolaczek — COMMENTED** [review]
Besides Alex's comment LGTM

File: setup.py (hunk 7 above)
Comment (mosheraboh) [reply setup.py]: Nice catch! Thanks.

**Comment by alex-golts** [comment]
@mosheraboh what currently prevents "fuse_examples" to be installed (or attempt to be installed) when someone does `pip install -e .[all]` (no examples mode)?
I don't see some sort of exclusion for it in setup.py

**Comment by mosheraboh** [comment]
Thanks @alex-golts!
Can you please pull and give it another try: `pip install -e .[all,examples]`

**Review by alex-golts — APPROVED** [review]

**Comment by alex-golts** [comment]
> Thanks @alex-golts!
> Can you please pull and give it another try: `pip install -e .[all,examples]`
> As for limiting a subpackage - there is no simple way (that I'm aware of) to exclude it. But the dependencies won't be installed - so you probably get an error when you try to use it.

works for me now, thanks

**Pull request merged**